File: SetupWizardMacApp/Sources/SetupWizard/Services/ProcessRunner.swift
import Foundation

/// Runs shell commands with real-time output streaming, cancellation support, and state tracking
actor ProcessRunner {
    enum RunnerError: LocalizedError {
        case processTerminated(exitCode: Int32)
        case cancelled
        case failedToStart
        case invalidExecutable
        
        var errorDescription: String? {
            switch self {
            case .processTerminated(let code):
                return "Process exited with code \(code)"
            case .cancelled:
                return "Process was cancelled by user"
            case .failedToStart:
                return "Failed to start process"
            case .invalidExecutable:
                return "Invalid or missing executable"
            }
        }
    }
    
    typealias OutputHandler = (String) -> Void
    
    private let executable: String
    private let arguments: [String]
    private var process: Process?
    private var outputLines: [String] = []
    private var isCancelled: Bool = false
    private var isRunning: Bool = false
    
    private var _onOutput: OutputHandler?
    
    func setOutputHandler(_ handler: @escaping OutputHandler) {
        _onOutput = handler
    }
    
    init(executable: String, arguments: [String] = []) {
        self.executable = executable
        self.arguments = arguments
    }
    
    /// Run the process with real-time output streaming
    func run(environment: [String: String] = [:]) async throws -> Int32 {
        let process = Process()
        self.process = process
        self.isCancelled = false
        self.isRunning = true
        self.outputLines = []
        
        // Verify executable exists
        guard FileManager.default.fileExists(atPath: executable) else {
            Logger.shared.error("Executable not found: \(executable)")
            self.isRunning = false
            throw RunnerError.invalidExecutable
        }
        
        process.executableURL = URL(fileURLWithPath: executable)
        process.arguments = arguments
        
        // Set up environment
        var env = ProcessInfo.processInfo.environment
        env.merge(environment) { _, new in new }
        process.environment = env
        
        // Set up pipes for output capture
        let stdoutPipe = Pipe()
        let stderrPipe = Pipe()
        process.standardOutput = stdoutPipe
        process.standardError = stderrPipe
        
        // Start streaming output (detached from actor)
        let outputQueue = DispatchQueue(label: "com.setupwizard.process-output")
        outputQueue.async { [weak self] in
            guard let self = self else { return }
            Task.detached { [weak self] in
                await self?.captureOutputAsync(from: stdoutPipe)
            }
        }
        outputQueue.async { [weak self] in
            guard let self = self else { return }
            Task.detached { [weak self] in
                await self?.captureOutputAsync(from: stderrPipe)
            }
        }
        
        // Start process
        do {
            try process.run()
            Logger.shared.debug("Started process: \(executable) with args: \(arguments)")
        } catch {
            Logger.shared.error("Failed to start process: \(error)")
            self.isRunning = false
            throw RunnerError.failedToStart
        }
        
        // Wait for completion
        process.waitUntilExit()
        
        // Give output capture threads a moment to process final data
        try? await Task.sleep(nanoseconds: 100_000_000) // 100ms
        
        if isCancelled {
            Logger.shared.log("Process cancelled")
            self.isRunning = false
            throw RunnerError.cancelled
        }
        
        let exitCode = process.terminationStatus
        self.isRunning = false
        
        if exitCode != 0 {
            Logger.shared.error("Process exited with code: \(exitCode)")
            throw RunnerError.processTerminated(exitCode: exitCode)
        }
        
        return exitCode
    }
    
    /// Cancel the running process gracefully
    func cancel() async {
        isCancelled = true
        
        guard let process = self.process, process.isRunning else {
            Logger.shared.log("No process to cancel")
            return
        }
        
        Logger.shared.log("Cancelling process (PID: \(process.processIdentifier))")
        
        // Send SIGTERM first (graceful)
        kill(process.processIdentifier, SIGTERM)
        
        // Wait up to 10 seconds for graceful termination
        let deadline = Date().addingTimeInterval(10)
        while process.isRunning && Date() < deadline {
            usleep(100_000) // 100ms
        }
        
        // If still running, send SIGKILL (forceful)
        if process.isRunning {
            Logger.shared.log("Force-killing process (PID: \(process.processIdentifier))")
            kill(process.processIdentifier, SIGKILL)
        }
    }
    
    private nonisolated func captureOutput(from pipe: Pipe) {
        let fileHandle = pipe.fileHandleForReading
        
        // Use readDataToEndOfFile() to block until EOF, not availableData which returns immediately
        let data = fileHandle.readDataToEndOfFile()
        
        guard !data.isEmpty, let string = String(data: data, encoding: .utf8) else {
            return
        }
        
        let lines = string.split(separator: "\n", omittingEmptySubsequences: false)
        for line in lines where !line.isEmpty {
            Task { await self.addOutputLine(String(line)) }
        }
    }
    
    private func captureOutputAsync(from pipe: Pipe) async {
        captureOutput(from: pipe)
    }
    
    private func addOutputLine(_ line: String) {
        outputLines.append(line)
        _onOutput?(line)
    }
    
    func getOutputLines() -> [String] {
        return outputLines
    }
    
    func isStillRunning() -> Bool {
        return isRunning && process?.isRunning ?? false
    }
}

File: SetupWizardCore/Services/ProcessRunner.cs
using System.Diagnostics;

namespace SetupWizardCore.Services;

/// <summary>
/// Runs shell commands with real-time output streaming, cancellation support, and state tracking.
/// Port of Swift ProcessRunner actor.
/// </summary>
public class ProcessRunner : IAsyncDisposable
{
    public class RunnerError : Exception
    {
        public int? ExitCode { get; }
        public RunnerError(string message, int? exitCode = null) : base(message)
        {
            ExitCode = exitCode;
        }
    }

    public delegate void OutputHandler(string line);

    private readonly string _executable;
    private readonly string[] _arguments;
    private Process? _process;
    private readonly List<string> _outputLines = new();
    private bool _isCancelled;
    private bool _isRunning;
    private OutputHandler? _onOutput;
    private readonly object _lock = new();
    private CancellationTokenSource? _cts;

    public ProcessRunner(string executable, string[]? arguments = null)
    {
        _executable = executable;
        _arguments = arguments ?? Array.Empty<string>();
    }

    public void SetOutputHandler(OutputHandler handler)
    {
        _onOutput = handler;
    }

    /// <summary>
    /// Run the process with real-time output streaming.
    /// Port of Swift run().
    /// </summary>
    public async Task<int> RunAsync(
        Dictionary<string, string>? environment = null,
        string? workingDirectory = null,
        CancellationToken cancellationToken = default)
    {
        _cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        _isCancelled = false;
        _isRunning = true;

        lock (_lock) _outputLines.Clear();

        // Verify executable exists
        if (!File.Exists(_executable))
        {
            SetupWizardLogger.Instance.Error($"Executable not found: {_executable}");
            _isRunning = false;
            throw new RunnerError("Invalid or missing executable");
        }

        var process = new Process();
        _process = process;

        process.StartInfo.FileName = _executable;
        process.StartInfo.Arguments = string.Join(" ", _arguments.Select(a => a.Contains(' ') ? $"\"{a}\"" : a));
        process.StartInfo.UseShellExecute = false;
        process.StartInfo.RedirectStandardOutput = true;
        process.StartInfo.RedirectStandardError = true;
        process.StartInfo.CreateNoWindow = true;

        // Set working directory
        if (!string.IsNullOrEmpty(workingDirectory))
            process.StartInfo.WorkingDirectory = workingDirectory;

        // Set environment variables
        if (environment != null)
        {
            foreach (var kvp in environment)
            {
                process.StartInfo.EnvironmentVariables[kvp.Key] = kvp.Value;
            }
        }

        // Set up output capture
        var outputCompletion = new TaskCompletionSource<bool>();
        var errorCompletion = new TaskCompletionSource<bool>();

        process.OutputDataReceived += (sender, args) =>
        {
            if (args.Data != null)
            {
                AddOutputLine(args.Data);
            }
        };

        process.ErrorDataReceived += (sender, args) =>
        {
            if (args.Data != null)
            {
                AddOutputLine(args.Data);
            }
        };

        process.Exited += (sender, args) =>
        {
            outputCompletion.TrySetResult(true);
            errorCompletion.TrySetResult(true);
        };

        try
        {
            if (!process.Start())
            {
                SetupWizardLogger.Instance.Error("Failed to start process");
                _isRunning = false;
                throw new RunnerError("Failed to start process");
            }

            SetupWizardLogger.Instance.Debug($"Started process: {_executable} with args: {string.Join(" ", _arguments)}");

            process.BeginOutputReadLine();
            process.BeginErrorReadLine();

            // Wait for exit or cancellation
            try
            {
                await process.WaitForExitAsync(_cts.Token);
            }
            catch (OperationCanceledException)
            {
                _isCancelled = true;
                await CancelAsync();
                _isRunning = false;
                throw new RunnerError("Process was cancelled by user");
            }

            // Ensure output has been fully read
            await Task.WhenAny(
                Task.WhenAll(outputCompletion.Task, errorCompletion.Task),
                Task.Delay(500));

            if (_isCancelled)
            {
                SetupWizardLogger.Instance.Log("Process cancelled");
                _isRunning = false;
                throw new RunnerError("Process was cancelled by user");
            }

            var exitCode = process.ExitCode;
            _isRunning = false;

            if (exitCode != 0)
            {
                SetupWizardLogger.Instance.Error($"Process exited with code: {exitCode}");
                throw new RunnerError($"Process exited with code {exitCode}", exitCode);
            }

            return exitCode;
        }
        catch (RunnerError) { throw; }
        catch (Exception ex)
        {
            SetupWizardLogger.Instance.Error($"Failed to start process: {ex.Message}");
            _isRunning = false;
            throw new RunnerError("Failed to start process");
        }
    }

    /// <summary>
    /// Cancel the running process gracefully (CloseMainWindow) then forcefully (Kill).
    /// Port of Swift cancel().
    /// </summary>
    public async Task CancelAsync()
    {
        _isCancelled = true;
        _cts?.Cancel();

        var process = _process;
        if (process == null || process.HasExited)
        {
            SetupWizardLogger.Instance.Log("No process to cancel");
            return;
        }

        SetupWizardLogger.Instance.Log($"Cancelling process (PID: {process.Id})");

        // Try graceful shutdown first
        if (!process.CloseMainWindow())
        {
            process.Kill(entireProcessTree: true);
        }
        else
        {
            // Wait up to 5 seconds for graceful termination
            try
            {
                await Task.Delay(5000, new CancellationTokenSource().Token);
                if (!process.HasExited)
                {
                    SetupWizardLogger.Instance.Log($"Force-killing process (PID: {process.Id})");
                    process.Kill(entireProcessTree: true);
                }
            }
            catch
            {
                if (!process.HasExited)
                {
                    process.Kill(entireProcessTree: true);
                }
            }
        }
    }

    public IReadOnlyList<string> GetOutputLines()
    {
        lock (_lock) return _outputLines.ToList().AsReadOnly();
    }

    public bool IsStillRunning()
    {
        return _isRunning && _process != null && !_process.HasExited;
    }

    private void AddOutputLine(string line)
    {
        lock (_lock) _outputLines.Add(line);
        _onOutput?.Invoke(line);
    }

    public async ValueTask DisposeAsync()
    {
        if (_process != null && !_process.HasExited)
        {
            _process.Kill(entireProcessTree: true);
        }
        _process?.Dispose();
        _cts?.Dispose();
        await Task.CompletedTask;
    }
}
